stages/code-traceability/src/main/java/edu/kit/kastel/mcse/ardoco/core/codetraceability/SamCodeTraceabilityLinkRecovery.java

```diff
@@ -26,7 +26,7 @@ public static SamCodeTraceabilityLinkRecovery get(SortedMap<String, String> addi
     protected void initializeState() {
         DataRepository dataRepository = getDataRepository();
         if (!DataRepositoryHelper.hasCodeTraceabilityState(dataRepository)) {
-            var codeTraceabilityState = new CodeTraceabilityStateImpl();
+            var codeTraceabilityState = new CodeTraceabilityStateImpl(dataRepository);
```

Comment:
Because of "global" structures. Static classes, singletons, etc are very bad for testing (especially in parallel). In the past there was a missing distinction between "actually global and stateless" and "global on a per-run basis". For example, WordSimUtils used to be a "static" class, however it is not stateless. In truth, it is a "global on a per-run basis" kind of class, that is shared between stages, but can be configured. 

So we have to find a solution to make such "global but not stateless" classes accessible to the stages without infringing on the separation between runs. The solution I came up with was attaching these classes to a thing that is already on a per-run basis -> the datarepository. Instead of referencing a static configurable class (bad), we can now make sure each run has its own instance in the datarepository, which it can configure as it pleases. However, since we can no longer just reference such a class statically, its instance has to be accessible to all constituents of the run. For example, the TextStateImpl used to statically refer to the sim utils, which it cannot do anymore. By having access to the datarepository, it can get the sim utils instance from it instead.

Ofcourse, in the particular case that you highlighted, the code traceability state currently doesn't make use of this. For an example of how this may be used, have a look at one of these stages

![grafik](https://github.com/ArDoCo/Core/assets/32802058/b43f10b5-d778-47d7-8a3a-dbdb9eb368bc)

I hope I explained it reasonably well, I suck at explaining

Reply:
mmh .. I get it. I'll think about it :)